tooling/minimize/src/constant.rs

```diff
@@ -2,87 +2,82 @@ use crate::*;
-            rs::mir::Const::Unevaluated(uneval, ty) =>
-                self.translate_const_uneval(uneval, *ty, span),
-            rs::mir::Const::Val(val, ty) => self.translate_const_val(val, *ty, span),
-        }
+        let val = c.eval(self.tcx, rs::ParamEnv::reveal_all(), None).expect("const-eval failed");
+        let tcx_at = self.tcx.at(span);
+        let (mut ecx, v) =
+            rs::mk_eval_cx_for_const_val(tcx_at, rs::ParamEnv::reveal_all(), val, c.ty()).unwrap();
+        self.translate_const_val(v, &mut ecx, span)
     }
 
     pub fn translate_const_smir(&mut self, c: &smir::Const, span: rs::Span) -> ValueExpr {
         self.translate_const(&smir::internal(self.tcx, c), span)
     }
 
-    fn translate_const_val(
+    fn translate_const_val<'mir>(
         &mut self,
-        val: &rs::ConstValue<'tcx>,
-        ty: rs::Ty<'tcx>,
+        val: rs::OpTy<'tcx>,
+        ecx: &mut rs::InterpCx<'mir, 'tcx, rs::CompileTimeInterpreter<'mir, 'tcx>>,
         span: rs::Span,
     ) -> ValueExpr {
-        let ty = self.translate_ty(ty, span);
-
-        let constant = match ty {
+        let ty = self.translate_ty(val.layout.ty, span);
+        match ty {
             Type::Int(int_ty) => {
-                let val = val.try_to_scalar_int().unwrap();
-                let int: Int = match int_ty.signed {
-                    Signed => val.try_to_int(val.size()).unwrap().into(),
-                    Unsigned => val.try_to_uint(val.size()).unwrap().into(),
+                let scalar = ecx.read_scalar(&val).unwrap();
+                let val = match int_ty.signed {
+                    Signed => scalar.to_int(scalar.size()).unwrap().into(),
+                    Unsigned => scalar.to_uint(scalar.size()).unwrap().into(),
                 };
-                Constant::Int(int)
+                ValueExpr::Constant(Constant::Int(val), ty)
             }
-            Type::Bool => Constant::Bool(val.try_to_bool().unwrap()),
-            Type::Tuple { fields, .. } if fields.is_empty() => {
-                return ValueExpr::Tuple(List::new(), ty);
+            Type::Bool => {
+                let val = ecx.read_scalar(&val).unwrap().to_bool().unwrap();
+                ValueExpr::Constant(Constant::Bool(val), ty)
             }
-            // A `static`
-            Type::Ptr(_) => {
-                let (alloc_id, offset) =
-                    val.try_to_scalar().unwrap().to_pointer(&self.tcx).unwrap().into_parts();
-                let alloc_id = alloc_id.expect("no alloc id?").alloc_id();
+            Type::Ptr(ptr_ty) => {
+                if let PtrType::FnPtr(_) = ptr_ty {
+                    rs::span_bug!(span, "Function pointers are currently not supported")
+                }
+                let ptr = ecx.read_pointer(&val).unwrap();
+                let (prov, offset) = ptr.into_parts();
+                let alloc_id = prov.unwrap().alloc_id();
```

Comment:
This can be `None` if the constant is something like
```rust
const C: *const i32 = 4 as *const i32;
```

Reply:
Now create a raw pointer if provenance is `None`.